### PR TITLE
Don't default to an invalid sqlite config if no database configuration is provided

### PR DESCRIPTION
**Pull request by nekatak**
Related to this ticket https://github.com/matrix-org/synapse/issues/6249

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by nekatak** [comment]
I wasnt aware of that I will check out and modify the MR. Thanks.

**Comment by nekatak** [comment]
@richvdh The requested changes should be ok now.

File: synapse/config/database.py [hunk 1]
```diff
@@ -62,7 +62,10 @@ def read_config(self, config, **kwargs):
         database_config = config.get("database")
 
         if database_config is None:
-            database_config = {"name": "sqlite3", "args": {}}
+            if config.get("database_path"):
+                database_config = {"name": "sqlite3", "args": {}}
+            else:
+                raise ConfigError("No database configuration provided")
```
[the remark this answers is not in the document]
Comment (nekatak) [reply synapse/config/database.py]: Updated the MR but didnt combine those methods. Instead I m running all the necessary checks in the read_arguments that is run after the read_config. Even like that it was a bit fiddly to catch the cases....

File: synapse/config/database.py [hunk 2]
```diff
@@ -127,27 +128,41 @@ def generate_config_section(self, data_dir_path, database_conf, **kwargs):
...
+                logger.warn("Ignoring 'database_path' for non-sqlite3 database")
+
+        else:
+            # no databases from the config file
+            if args.database_path is None:
+                raise ConfigError("No database config provided")
+            else:
+                database_config = {"name": "sqlite3", "args": {}}
+                self.databases = [DatabaseConnectionConfig("master", database_config)]
+                self.set_databasepath(args.database_path)
```
Comment (richvdh) [inline synapse/config/database.py]: I think the flow in this function would be easier to follow if you dealt with the case of `args.database_path is None` first:

```
if args.database_path is None:
    if not databases:
        raise ConfigError...
   return
```

generally I prefer early `returns` to nested `if/else` statements: they functions easier to follow.

**Comment by nekatak** [comment]
Thanks for your feedback! I ll make the changes and update the MR.

**Comment by richvdh** [comment]
Please don't force-push branches during review: it makes it hard for me to see what's changed since last time I looked :/

**Comment by nekatak** [comment]
> Please don't force-push branches during review: it makes it hard for me to see what's changed since last time I looked :/

Sorry. I wasnt aware that github is not keeping clear versioning of the MRs. I was trying to keep a single commit message.

File: synapse/config/database.py [hunk 3]
```diff
@@ -76,26 +81,27 @@ def read_config(self, config, **kwargs):
...
-                database_config = {"name": "sqlite3", "args": {}}
-
+        elif database_config:
             self.databases = [DatabaseConnectionConfig("master", database_config)]
 
-            self.set_databasepath(config.get("database_path"))
+        elif database_path:
+            database_config = {"name": "sqlite3", "args": {}}
+            self.databases = [DatabaseConnectionConfig("master", database_config)]
+            self.set_databasepath(database_path)
```
[the remark this answers is not in the document]
Comment (nekatak) [reply synapse/config/database.py]: I think as it is now keeps the behavior it had before.

**Pull request merged**